lib/metadata/bucketfile/backend.js

```diff
@@ -0,0 +1,520 @@
+        } else {
+            this.refcnt = 0;
+            this.waitreco = 0;
+            this.realReConnect();
+            this.notifier = new events.EventEmitter();
+        }
+    }
+
+    /**
+     * Start the server
+     * @return {undefined}
+     */
+    startServer() {
+        const rootDB = level(METADATA_PATH + ROOT_DB);
+        const sub = sublevel(rootDB);
+        sub.methods = sub.methods || {};
+        sub.methods.createSub = { type: 'async' };
+        sub.createSub = (subName, cb) => {
+            try {
+                sub.sublevel(subName);
+                multilevel.writeManifest(sub,
+                                         METADATA_PATH +
+                                         MANIFEST_JSON_TMP);
+                fs.renameSync(METADATA_PATH + MANIFEST_JSON_TMP,
+                              METADATA_PATH + MANIFEST_JSON);
+                cb();
+            } catch (err) {
+                cb(err);
+            }
+        };
+        const metastore = sub.sublevel(METASTORE);
+        /* Since the bucket creation API is expecting the
+           usersBucket to have attributes, we pre-create the
+           usersBucket here */
+        sub.sublevel(constants.usersBucket);
+        const usersBucketAttr = new BucketInfo(constants.usersBucket,
+                                               'admin', 'admin',
+                                               new Date().toJSON(), 2);
+        metastore.put(constants.usersBucket, usersBucketAttr.serialize());
+        const stream = metastore.createKeyStream();
+        stream
+            .on('data', e => {
+                // automatically recreate existing sublevels
+                sub.sublevel(e);
+            })
+            .on('error', err => {
+                this.logger.fatal('error listing metastore', { error: err });
+                throw (errors.InternalError);
+            })
+            .on('end', () => {
+                multilevel.writeManifest(sub, METADATA_PATH + MANIFEST_JSON);
+                this.logger.info('starting metadata file backend server');
+                /* We start a server that will server the sublevel
+                   capable rootDB to clients */
+                net.createServer(con => {
+                    con.pipe(multilevel.server(sub)).pipe(con);
+                }).listen(METADATA_PORT);
+            });
+    }
+
+    /**
+     * Reconnect to the server
+     * @return {undefined}
+     */
+    realReConnect() {
+        if (this.client !== undefined) {
+            this.client.close();
+        }
+        delete require.cache[require.resolve(METADATA_PATH + MANIFEST_JSON)];
+        const manifest = require(METADATA_PATH + MANIFEST_JSON);
+        this.client = multilevel.client(manifest);
+        const con = net.connect(METADATA_PORT);
+        con.pipe(this.client.createRpcStream()).pipe(con);
+        this.metastore = this.client.sublevel(METASTORE);
+    }
+
+    /**
+     * Wait for reconnect do be done
```

Comment:
the comment should be: try to reconnect possibly waiting for other operations to complete